"""
Defines the help text for the QIIME pipeline forms.
"""
from nephele2 import tfvars

qiime_base_phred_qual_desc = """
    Phred quality score is a measure of the quality of the identification of the nucleobases generated by
    the sequencing platforms such as Illumina and 454. Raw reads generated by Illumina or 454 have
    predicted error probabilities for each base. Phred quality scores <em>Q</em> are defined as a
    property which is logarithmically related to the base calling error probabilities <em>P</em>. Minimum
    Phred quality score of 19 for Q20 (1 in 100 of incorrect base call) or better is recommended.
    """

qiime_base_phred_offset_desc = """
    Phred <em>Q</em> scores are often represented as ASCII characters of base 33 and 64. Base 33 is the
    most common representation on modern sequencing platforms, while 64 on 454 and older Illumina. The
    default is base 33.
    """

qiime_base_database_desc = """
    <strong>HOMD</strong>: the HOMD database will be used for taxonomic assignment. Note: HOMD (Human Oral Microbiome Database) is a custom database for the analysis of human oral microbiome only.</br>
    <strong>SILVA</strong>: the SILVA database will be used for taxonomic assignment.<br>
    <strong>Greengenes</strong>: the Greengenes database will be used for taxonomic assignment.<br>
    <strong>99</strong>: sequences were clustered at 99% identity.<br>
    <strong>97</strong>: sequences were clustered at 97% identity.
    """

qiime_picrust_desc = """
    Checking this box to run <a href='http://picrust.github.io/picrust/' target="_blank" rel="noopener noreferrer">PICRUSt</a>.
    PICRUSt predicts metagenome functional content from marker genes. The analysis will
    produce two functional bar charts at taxa level 2 (Phylum) and 3 (Class).<br>
    Note: PICRUSt runs separately using Closed Reference OTU picking with the Greengenes 99 database.
    """

qiime_base_analysis_type_desc = """
    <strong>Closed Reference</strong>: reads are mapped against a database of reference
    sequences and reads that do not map to the reference sequences are excluded from all
    downstream analyses.<br>
    <strong>Open Reference</strong>: reads are mapped against a database of reference
    sequences and reads that do not map to the reference sequences are subsequently
    clustered <em>de novo</em>.<br>
    <strong><em>de novo</em></strong>: reads are clustered against each another without
    any external reference sequences.<br>
    See <a href='http://qiime.org/tutorials/otu_picking.html' target='_blank' rel="noopener noreferrer">QIIME</a>
    for more information.
    """

sampling_depth_desc = "The number of counts for filtering and subsampling the OTU table for downstream analysis. \
    Samples which have counts below this value will be removed from the downstream analysis. \
    The counts of the remaining samples will be subsampled to this value. If not specified, \
    it will be calculated automatically. \
    <a href=\""+tfvars.SERVER_ADDR+"/faq/#collapse12\" target=\"_blank\">See FAQ: Why are only some of \
    my samples appearing in the downstream analysis for the amplicon pipelines?</a>"

qiime_reverse_compliment_desc = """
    Enable reverse strand match for OTU picking. If selected, reads will map in forward
    and reverse direction. If not selected, reads will only map in the forward direction.
    This parameter should be enabled if reads are stored as reverse complements.
    """

qiime_chimera_removal_desc = """
    Select this option if you want to remove chimeric sequences after OTU picking.
    Chimera removal is carried out by <code>vsearch</code>
    (<a href='https://peerj.com/articles/2584/' target="_blank" rel="noopener noreferrer">T. Rognes</a>).<br>
    Note: this option does not apply to Closed Reference workflow.
    """

qiime_fraction_max_sample_size_desc = """
    MUST BE GREATER THAN 0.<br>
    Minimum sample size for the Core Diversity Analysis; specify as an integer (>10)
    or a fraction of the median sample in the dataset (a decimal between 0 and 1).<br>
    Minimum number of reads: minimum sample size to include in the downstream
    quantitative analysis; specify an integer (>10). Samples included in analysis will
    be rarefied to this size.<br>
    Fraction of median sample size: number between 0 and 1. The minimum sample size
    is computed from this fraction of the median sample size. Samples will be rarefied
    based on the number reads in the smallest sample which exceeds this minimum size.<br>
    For example, if you have four samples A-D in your dataset, with read counts of 200,
    1500, 2000, and 2500, respectively, specifying 0.2 will remove sample A because it
    has fewer than 350 reads (i.e., 20% of the median which is 350 reads) and the
    remaining samples will be randomly subsampled to the lowest of the remaining samples,
    in this case 1500. If instead you provide an integer such as 2000, the samples with
    read counts 200 and 1500 will be removed and all remaining samples will be subsampled
    at the integer you indicated (2000).
    """

qiime_core_microbiome_desc = """
    Calculates the shared microbiome within samples of the same TreatmentGroup.<br>
    NOTE: the analysis may fail if the samples are considerably diverse.
    """

qiime_pe_max_ambiguous_desc = """
    Maximum number of degenerate bases permitted in a sequence; this is applied
    after quality trimming, and is total over combined paired end reads if
    applicable.
    """

qiime_max_bad_run_length_desc = """
    Maximum number of consecutive low quality base calls permitted before
    truncation. For V1-V3, a minimum value of 10 is recommended.
    """

qiime_min_overlap_desc = """
    Minimum number of overlapped bases for the join of paired-end reads. Must be
    an integer.
    """

qiime_percent_max_diff_desc = """
    Maximum percentage of differences in the overlapped regions. Must be an
    integer between 1 and 100.
    """

qiime_min_seq_length_desc = """
    Minimum sequence length, in bases.
    """

qiime_max_seq_length_desc = """
    Maximum sequence length, in bases.
	"""

qiime_min_quality_score_desc = """
    Minimum average quality score in a read.
	"""

qiime_max_ambiguos_desc = """
    Maximum number of degenerate bases in a read.
	"""

qiime_max_homopolymer_desc = """
    Maximum length of homopolymer run within a read.
	"""

qiime_max_primer_mismatch_desc = """
    Maximum number of primer mismatches.
	"""

qiime_max_barcode_errs_desc = """
    Maximum number of mismatches permitted in a barcode.
	"""

qiime_quality_score_window_desc = """
    Enable sliding window test of quality scores. If the average score of a
    continuous set of value <strong><em>w</em></strong> bases falls below the
    threshold (see <strong>Minimum qual score</strong> for default), the
    sequence is discarded. A recommended value for window width is
    <strong>50</strong>. A value of <strong>0</strong> (zero) implies no filtering.
    Default behavior for this function is to truncate the sequence at the beginning
    of the poor quality window, and test for minimal length (<strong>Minimum
    sequence length</strong> parameter) of the resulting sequence.
    """

qiime_disable_barcode_correction_desc = """
    Disable any attempt to find nearest corrected barcode.
	"""

qiime_primer_correcting_formats_desc = """
    <strong>Barcode length from mapping file</strong>: barcode length will be determined from the mapping file and a normal barcode correction method is activated (See
    <a href='http://qiime.org/scripts/split_libraries.html' target='_blank' rel="noopener noreferrer">QIIME</a> for more information).<br>
    <strong>Golay 12</strong>: 12 bp error-correcting barcode sequences based on the Golay class of codes. See <a href='http://www.pnas.org/content/105/46/17994.full' target="_blank" rel="noopener noreferrer">PNAS 105 (46): 17994-9 (2008)</a> for more information.<br>
    <strong>Hamming 8</strong>: 8 bp error-correcting barcode sequences based on the Hamming class of codes. See <a href='https://www.nature.com/articles/nmeth.1184' target="_blank" rel="noopener noreferrer">Nature Methods 5, 235-237 (2008)</a> for more information.<br>
    <strong>Variable Length</strong>: Barcode length will be variable.
    """

qiime_its_max_bad_run_length_desc = """
    Max number of consecutive low quality bases permitted before truncation.
    """

qiime_its_ref_db_desc = """
    <a href='https://unite.ut.ee/' target="_blank" rel="noopener noreferrer">UNITE</a> is a rDNA sequence database designed to provide a stable and reliable platform for sequence-borne identification of ectomycorrhizal asco- and basidiomycetes.<br>
    <a href="https://nephele.niaid.nih.gov/details_qiime/">Which version we use?</a><br>
    <strong>ITS 99</strong>: sequences were clustered at 99% identity.<br>
    <strong>ITS 97</strong>: sequences were clustered at 97% identity.
    """
