{% extends "base.html" %}

{% block content %}
  <div>
  <!-- <div style="padding: 20px 30px 20px 30px"> -->
    <h2>Taxonomic Assignment</h2>
    <p>Taxonomic assignment include using a <em>k</em>-Nearest Neighbor consensus and <a href="http://aem.asm.org/content/73/16/5261" target="_blank" rel="noopener noreferrer">Wang approach</a>. The <em>k</em>-Nearest Neighbor algorithm involves identifying the <em>k</em>-most similar sequences in a database that are similar to your sequence. By default, mothur will find the 10 most similar sequences in the database. Once mothur has identified the <em>k</em>-most similar sequences, it will use the taxonomy information for each sequence to determine the consensus taxonomy. mothur gives you the ability to determine the method that is used to find the closest matches. The Wang method looks at the query sequence <em>k</em>mer by <em>k</em>mer. The method looks at all taxonomies represented in the template, and calculates the probability a sequence from a given taxonomy would contain a specific <em>k</em>mer. Then calculates the probability a query sequence would be in a given taxonomy based on the <em>k</em>mers it contains, and assign the query sequence to the taxonomy with the highest probability. This method also runs a bootstrapping algorithm to find the confidence limit of the assignment by randomly choosing with replacement 1/8 of the <em>k</em>mers in the query and then finding the taxonomy.</p>
    <h2>Alpha Diversity</h2>
    <p>Alpha diversity refers to the diversity within a particular ecosystem, and is usually expressed by the number of species or OTUs in that ecosystem. As an example, a study might be monitoring how treatment affects the diversity of the indigenous members/OTUs in an ecosystem. The calculation of alpha diversity is influenced by the sequencing depth. Higher sequencing depth generallhy yield more diversity. In addition, definitions of alpha diversity differ on the assumptions of diversity. The values given by one or more diversity indices, such as species richness, the Shannon index or the Simpson index.</p>
    <h2>Rarefaction Curve</h2>
    <p>Rarefaction is a technique to assess species richness from the results of sampling. It allows the calculation of species richness for a specified sequencing depth, based on the construction of rarefaction curves. This curve is a plot of the number of species or OTUs as a function of the number of samples. The goal of rarefaction is to get an indication of whether enough observations have been made to get a good measurement of an alpha diversity metric. Rarefaction curves generally grow rapidly at first, as the most common OTUs are found, but the curves plateau as only the rarest OTUs remain to be sampled. Sampling various OTUs in a community is correlated to the sequencing depth. More OTUs will be found with higher sequencing depth. Rarefaction curves are created by randomly resampling the pool of <em>N</em> samples multiple times and then plotting the average number of OTUs found in each sample. Rarefaction works well when no OTUs are extremely rare or common.</p>
    <h2>Beta Diversity</h2>
    <p>Beta diversity measures the change in diversity of species from one sample to another. It calculates the number of species that are not the same in two different samples. A high beta diversity index indicates a low level of similarity, while a low beta diversity index shows a high level of similarity. <a href="https://www.nature.com/articles/ismej2010133" target="_blank" rel="noopener noreferrer">UniFrac</a>, as an example, is a distance metric that incorporates information on the relative relatedness of community members by incorporating phylogenetic distances between observed OTUs in the computation. Weighted UniFrac accounts for abundance of observed OTUs while unweighted UniFrac only considers the presence and absence of OTUs in the communities. Weighted UniFrac is more appropriate to reveal community differences that are due to changes in relative OTU abundance. Unweighted UniFrac, on the other hand, is most informative when communities differ primarily by what can live in them, in part because abundance information can obscure significant patterns of variation in which OTUs are present.</p>
    <h2>Principal Coordinate Analysis (PCoA)</h2>
    <p>Principal Coordinate Analysis (PCoA) is a multivariate statistical method to explore and to visualize similarities or dissimilarities of data. It is commonly used to compare groups of samples based on phylogenetic or count-based distance metrics. PCoA summarizes and attempts to represent sample similarities and dissimilarities in a low-dimensional, Euclidean space. Computationally, PCoA produces a set of uncorrelated axes to summarise the variability in the data set. A successful PCoA will generate a few axes with relatively large eigenvalues, capturing above 50 percent of the variation in the samples, with all other axes having small eigenvalues. Samples ordinated closer to one another are more similar than those ordinated further away. It is worth noting that while PCoA is suited to handling a wide range of data, information concerning the original variables cannot be recovered. This is because PCoA takes a similarity/dissimilarity matrix derived from the original data as input and not the original variables themselves.</p>
</div>
{% endblock %}