{% extends "base.html" %}

{% block page_content %}
  <!--<div style="padding: 20px 30px 20px 30px">-->
<div style="padding: 20px 30px 20px 30px">
    <h1>QIIME1 Pipeline</h1>
    <p>QIIME1 is an open-source bioinformatics pipeline for performing microbiome analysis from raw DNA sequencing data. QIIME1 is designed to take users from raw sequencing data generated on the Illumina or other platforms through publication quality graphics and statistics. This includes demultiplexing and quality filtering, OTU picking, taxonomic assignment, and phylogenetic reconstruction, and diversity analyses and visualizations. QIIME1 has been applied to studies based on billions of sequences from tens of thousands of samples.</p>
    <h2>User Options</h2>
    <h3>Preprocessing</h3>
    <ul>
        <li>Minimum Phred quality score: The Phred quality score is a measure of the quality of the identification of the nucleobases generated by the sequencing platforms such as Illumina and 454. Minimum Phred quality score of 19 for Q20 (1 in 100 of incorrect base call) or better is recommended. The default is 19.</li>
        <li>Phred offset: Phred <em>Q</em> scores are often represented as ASCII characters of base 33 and 64. Base 33 is the most common representation on modern sequencing platforms, while 64 on 454 and older Illumina. The default is base 33.</li>
        <li>Maximum ambiguous: Maximum number of degenerate bases (N) allowed in a sequence to retain it. This is applied after quality trimming, and is total over combined paired end reads if applicable. The default is 0.</li>
    </ul>
    <h3>Join Reads</h3>
    <ul>
        <li>Max bad run length: Maximum number of consecutive low quality base calls permitted before truncation. For V1-V3, a minimum value of 10 is recommended. The default is 3.</li>
        <li>Minimum overlap length: Minimum number of overlapped bases for the join of paired-end reads. Must be an integer. The default is 10.</li>
        <li>Percent difference with overlap: Maximum percentage of differences in the overlapped regions. Must be an integer between 1 and 100. The default is 25 or 25%.</li>
    </ul>
    <h3>Analysis</h3>
    <ul>
        <li>Sampling depth: The number of counts for filtering and subsampling the OTU table for downstream analysis. Samples which have counts below this value will be removed from the downstream analysis. The counts of the remaining samples will be subsampled to this value. If not specified, it will be calculated automatically. (See FAQ: <a href="{{ url_for('show_FAQ', _anchor='collapse12') }}">Why are only some of my samples appearing in the downstream analysis for the amplicon pipelines?</a>)</li>
    </ul>
    <h2>Taxonomic Assignment</h2>
    <p>All of the sequences from each sample will be clustered into Operational Taxonomic Units (OTUs) based on their sequence similarity. There are seven databases to choose for taxonomic assignment, namely <a href="http://www.homd.org/" target="_blank">HOMD (Human Oral Microbiome Database; v15.1)</a>, SILVA97 and SILVA99 (v132), Greengenes 97 and Greengenes 99 (v13.8), and ITS97 and ITS99 (<a href="https://unite.ut.ee" target="_blank">UNITE+INSDC</a> 18.11.2018). The ITS databases have been updated to the latest version with 817,130 sequences. The sequences in the UNITE databases target the formal fungal barcode, the nuclear ribosomal internal transcribed spacer (ITS) region. UNITE uses the NCBI taxonomy classification with modifications from <a href="http://www.indexfungorum.org" target="_blank">Index Fungorum</a>. The sequences in these databases are clustered at 97 and 99% identity. There are three main strategies for OTU picking:</p>
    <ul>
        <li><b><em>De Novo</em></b>: Reads are clustered against one another without an external reference sequence collection. Useful for studying populations where there is poor characterization of existing data. In addition to clustering, the pipeline also performs taxonomy assignment, sequence alignment, and tree-building steps.</li>
        <li><b>Closed reference</b>: Reads are clustered against a reference sequence collection and any reads which do not hit a sequence in the reference sequence collection are excluded. Useful when existing taxa are well characterized and one is not interested in novel ones. Taxonomic assignments in the reference database are assigned to the OTUs.</li>
        <li><b>Open reference</b>: Reads are clustered against a reference sequence collection and any reads which do not hit the reference collection are clustered <em>de novo</em>. Applicable when existing taxa are well characterized and one also wishes to discover novel ones.</li>
    </ul>
    <h2>Pipeline Major Steps</h2>
    <ol>
        <li>Join forward and reverse short reads as contigs: Contigs are generated using the command <code>join_paired_ends.py</code>. Several parameters are used to determine the overlap and quality of reads to create a contig.
            <ul>
                <li><code>pe_join_method</code>: method to use for joining paired-ends, default fastq-join.</li>
                <li><code>min_overlap</code>: minimum allowed overlap in base-pairs required to join pairs, default 10 (must be integer).</li>
                <li><code>perc_max_diff</code>: maximum allowed percentage of differences within region of overlap, default 25 (between 1-100; only applied to fastq-join method).</li>
            </ul>
        </li>
        <li>Demultiplex FASTQ sequences: The <code>split_libraries_fastq.py</code> command performs demultiplexing of FASTQ sequence data.
            <ul>
                <li><code>max_bad_run_length</code>: max number of consecutive low quality base calls allowed before truncating a read, default 3.</li>
                <li><code>sequence_max_n</code>: maximum number of degenerate bases (N) allowed in a sequence to retain it. This is applied after quality trimming, and is total over combined paired end reads if applicable, default 0.</li>
                <li><code>phred_quality_threshold</code>: default 19.</li>
                <li><code>phred_offset</code>: the ASCII offset to use when decoding Phred scores (either 33 or 64), default 33.</li>
            </ul>
        </li>
        <li>OTU picking: The OTU picking step, <code>pick_open_reference_otus.py</code>, <code>pick_closed_reference_otus.py</code>, or <code>pick_de_novo_otus.py</code>, assigns similar sequences to operational taxonomic units, or OTUs, by clustering sequences based on a user-defined similarity threshold. Sequences which are similar at or above the threshold level are taken to represent the presence of a taxonomic unit (e.g., a genus, when the similarity threshold is set at 0.94) in the sequence collection.
        </li>
        <li>Analysis of core diversity: The <code>core_diversity_analyses.py</code> is a workflow for running a core set of QIIME1 diversity analyses, beginning with a BIOM table, mapping file, and optional phylogenetic tree. The commands include <code>alpha_rarefaction.py</code>, <code>beta_diversity_through_plots.py</code>, <code>summarize_taxa_through_plots.py</code>, plus the (non-workflow) scripts <code>make_distance_boxplots.py</code>, <code>compare_alpha_diversity.py</code>, and <code>group_significance.py</code>.
        </li>
        <li>PICRUSt (Optional): PICRUSt is designed to estimate the gene families contributed to a metagenome by bacteria or archaea identified using 16S rRNA sequencing. Intermediate steps in this pipeline may also be of independent interest, as they allow for phylogenetic prediction of organismal traits using reference examples (here applied to the problem of gene content prediction), and correction for variable marker gene copy number.
        </li>
    </ol>
    <h2 id="output-files">Output Files/Directories</h2>
    <ul>
        <li><b>split_lib_out</b>: Demultiplex and quality filter (usually at Phred > Q20) of paired Illumina reads. For more information, please visit <a href="http://qiime.org/scripts/split_libraries_fastq.html" target="_blank" rel="noopener noreferrer">QIIME1 online documentation</a>.</li>
        <li><b>otus</b>: The OTU picking step assigns similar sequences to operational taxonomic units, or OTUs, by clustering sequences based on a user-defined similarity threshold. Sequences which are similar at or above the threshold level are taken to represent the presence of a taxonomic unit (e.g., a genus, when the similarity threshold is set at 0.94) in the sequence collection. The final OTU table is summarized in a <a href="http://biom-format.org/" target="_blank">BIOM file</a>, e.g., <code>otu_table_mc2_w_tax_no_pynast_failures.biom</code> for open reference OTU picking strategy and <code>otu_table.biom</code> for closed and <i>de novo</i>. These BIOM files are used for the downstream analysis. For more information, please visit <a href="http://qiime.org/scripts/pick_otus.html" target="_blank" rel="noopener noreferrer">QIIME1 online documentation</a>.</li>
        <li><b>core_dirersity</b>: The diversity analysis generate figures and tables according to the metadata in the mapping file. The analysis includes alpha diversity, beta diversity, distance boxplots, comparison of alpha diversity, and quantitative analysis of group differences. The output from core diversity is conveniently summarized in a HTML file (<code>index.html</code>). The HTML file lists all the quantitative analyses such as alpha and beta diversity, group significance, and the corresponding figures and summaries. For more information, please visit <a href="http://qiime.org/scripts/core_diversity_analyses.html" target="_blank" rel="noopener noreferrer">QIIME1 online documentation</a>.<br><strong>Note:</strong> The QIIME1 pipeline no longer generates the taxonomy summary plots (<code>plot_taxa_summary.py</code>). For microbial communities with complex profiles, it can take a few weeks to generate the taxonomy summary plots. Please consider the <a href="https://nephele.niaid.nih.gov/upload_biom" target="_blank">Downstream Analysis pipeline</a> for additional analysis.</li>
        <li><b>graphs</b>: The visualization pipeline runs the <a href="https://plot.ly/python/" target="_blank" rel="noopener noreferrer">plotly graphs</a> and generates an interactive heatmap. Please visit the <a href="{{ url_for('show_tutorials') }}">16S visualization tutorial</a> page to learn how to use it. For more information about what tools we used, please visit the <a href="{{ url_for('show_vis_details') }}">16S visualization pipeline</a> page.</li>
        <li><code>otu_summary_table.txt</code>: The file lists the number of taxonomically assigned reads for each sample in the dataset.</li>
        <li><code>otu_table.txt</code>: A tab-separated text file of sequence variant counts and taxonomic assignment at the genus level.</li>
        <li><b>otu_picrust</b> (optional): Closed reference OTU picking using QIIME1 with Greengenes 99 database.</li>
        <li><b>PICRUST_data</b> (optional): This directory contains the final metagenome functional predictions with figures and bar plots. For more information, please visit <a href="http://picrust.github.io/picrust/" target="_blank" rel="noopener noreferrer">PICRUSt online documentation.</a></li>
    </ul>
    <h2 id="tools-and-references">Tools and References</h2>
    <ol>
        <li>JG Caporaso, J Kuczynski, J Stombaugh, K Bittinger, FD Bushman, EK Costello, N Fierer, A Gonzalez Pena, JK Goodrich, JI Gordon, GA Huttley, ST Kelley, D Knights, JE Koenig, RE Ley, CA Lozupone, D McDonald, BD Muegge, M Pirrung, J Reeder, JR Sevinsky, PJ Turnbaugh, WA Walters, J Widmann, T Yatsunenko, J Zaneveld and R Knight (2010) QIIME allows analysis of high-throughput community sequencing data. Nature Methods, doi: <a href="https://www.nature.com/articles/nmeth.f.303" target="_blank" rel="noopener noreferrer"> 10.1038/nmeth.f.303</a>.</li>
        <li>MGI Langille, J Zaneveld, JG Caporaso, D McDonald, D Knights, JA Reyes, JC Clemente, DE Burkepile, RL Vega Thurber, R Knight, RG Beiko and C Huttenhower (2013) Predictive functional profiling of microbial communities using 16S rRNA marker gene sequences. Nature Biotechnology, doi: <a href="https://www.nature.com/articles/nbt.2676" target="_blank" rel="noopener noreferrer">10.1038/nbt.2676</a>.</li>
        <li>RH Nilsson, KH Larsson, AFS Taylor, J Bengtsson-Palme, TS Jeppesen, D Schigel, P Kennedy, K Picard, FO Glöckner, L Tedersoo, I Saar, U Kõljalg, K Abarenkov (2019) The UNITE database for molecular identification of fungi: handling dark taxa and parallel taxonomic classifications. Nucleic Acids Research, doi: <a href="https://doi.org/10.1093/nar/gky1022">10.1093/nar/gky1022</a>.</li>
        <li>UNITE Community (2019): UNITE QIIME release for Fungi. Version 18.11.2018. UNITE Community. <a href="https://doi.org/10.15156/BIO/786334">https://doi.org/10.15156/BIO/786334</a>.</li>
    </ol>
  </div>
{% endblock %}
