# Synopsis #
QIIME (Quantitative Insights Into Microbial Ecology) is a bioinformatic pipeline designated for the task of analyzing
microbial communities that were sampled through marker gene (such as 16S or 18S rRNA genes) amplicon sequencing. It is
designed to take users from raw sequencing data generated on the platform, such as Illumina, to publication quality
graphics and statistics. QIIME includes functions such as demultiplexing and quality filtering, OTU picking (open,
closed, and *de novo*), taxonomic assignment, and phylogenetic reconstruction, and diversity analyses and visualizations.
The main output from the QIIME pipeline is a feature table, which describes the abundances of each OTU or sequence
variant in each sample. Additional tools that are relevant to ecological aspects of the samples being investigated are
also provided within the pipeline, including rarefaction, alpha diversity and beta diversity calculations, visualizations
such as principle coordinates analysis (PCoA) and much more. The QIIME pipelines include an option that runs
[PICRUSt](https://github.com/picrust/picrust2/wiki). PICRUSt (Phylogenetic Investigation of Communities by
Reconstruction of Unobserved States) is a bioinformatics software package designed to predict metagenome functional
content from marker gene (e.g., 16S rRNA) surveys and full genomes. The QIIME pipelines currently support the
[Human Oral Microbiome Database (HOMD)](https://www.homd.org), [SILVA 97/99](https://www.arb-silva.de) and
[Greengenes 97/99 databases](https://greengenes.secondgenome.com).

# Example Usage #
The QIIME pipelines for paired-end, single-end, and ITS samples share the same code base. Therefore, the options for
each pipeline is quite similar.

## QIIME Paired End ##
```bash
qiime_pe.py --job_id job_id --data_type PE --map_file map_file.txt --min_overlap 10 --perc_max_diff 25 --phred_quality 19 --phred_offset 33 --max_bad_run 3 --max_n 0 -sampling_depth 1500 --otu_strategy open --ref_db gg99 --picrust
```

## QIIME Single End ##

```bash
qiime_se.py --job_id job_id --data_type SE --map_file map_file.txt --phred_quality 19 --phred_offset 33 --max_bad_run 3 --max_n 0 -sampling_depth 1500 --otu_strategy open --ref_db gg99 --picrust
```

## QIIME ITS ##
```bash
qiime_its.py --job_id job_id --data_type ITS --map_file map_file.txt --min_overlap 10 --perc_max_diff 25 --phred_quality 19 --phred_offset 33 --max_bad_run 3 --max_n 0 -sampling_depth 1500 --otu_strategy open --ref_db its97 --picrust
```

## Pipeline Options ##
- ``--job_id``: job id.
- ``--data_type``: analysis data type. The options are ``PE``, ``SE``, and ``ITS``.
- ``--map_file``: mapping file that describes the metadata associated with samples. The mapping file should contain all of
the information about the samples necessary to perform the data analysis. In general, the mapping file should contain the
name of each sample, the barcode sequence used for each sample, the linker/primer sequence used to amplify the sample, and
a description column. One should also include in the mapping file any metadata that relates to the samples (for instance,
health status or sampling site) and any additional information relating to specific samples that may be useful to have at
hand when considering outliers (for example, what medications a patient was taking at time of sampling). See
[QIIME documentation](http://qiime.org/documentation/file_formats.html) for the format.
- ``--min_overlap``: minimum allowed overlap in bases required to join pairs. Not applicable to single-end samples. The
default value is 10.
- ``--perc_max_diff``: maximum allowed percent differences within overlap. Not applicable to single-end samples. The default
value is 25.
- ``--phred_quality``: maximum unacceptable [Phred quality score](https://www.phrap.com/phred/). The default value is 19.
The Phred quality score is a measure of the quality of the identification of the nucleobases generated by the sequencing
platforms such as Illumina and 454. Minimum Phred quality score of 19 for Q20 (1 in 100 of incorrect base call) or better is
recommended. The default is 19.
- ``--phred_offset``: the ASCII offset to use when decoding [Phred scores]((https://www.phrap.com/phred/)) (either 33 or 64).
The default is 33. Phred Q scores are often represented as ASCII characters of base 33 and 64. Base 33 is the most common
representation on modern sequencing platforms, while 64 on 454 and older Illumina. The default is base 33.
- ``--max_bad_run``: maximum number of consecutive low quality base calls permitted before truncation. For V1-V3, a minimum
value of 10 is recommended. The default value is 3.
- ``--max_n``: maximum number of degenerate bases (N). The default value is 0.
- ``--sampling_depth``: sampling depth for downstream analysis. The number of counts for filtering and subsampling the OTU
table for downstream analysis. Samples which have counts below this value will be removed from the downstream analysis. The
counts of the remaining samples will be subsampled to this value. If not specified, it will be calculated automatically. The
default value is ``None``.
- ``--otu_strategy``: OTU picking strategy. The options are ``open``, ``closed``, and ``de_novo``.
- ``--ref_db``: reference database for OTU picking. The options are ``homd`` (HOMD), ``sv97`` (SILVA with 97% identity),
``sv99`` (99%), ``gg97`` (97%), and ``gg99`` (99%).
- ``--picrust``: run [PICRUSt](https://github.com/picrust/picrust2/wiki). PICRUSt (Phylogenetic Investigation of Communities
by Reconstruction of Unobserved States) is a software for predicting functional abundances based only on marker gene
sequences.

## QIIME Parameter Files ##
The QIIME workflow scripts combine two or more QIIME commands to facilitate running common processes. For example,
``pick_de_novo_otus.py`` combines OTU picking, representative sequence picking, taxonomy assignment, sequence alignment,
alignment filtering, and tree building in a single QIIME command. Because each of these individual commands may have many
options, it is generally not feasible to make of these options accessible through the ``pick_de_novo_otus.py`` interface.
Therefore, to simplify the execution of the analysis, it is advised to use the parameter file to control different options in
the indivudal scripts. To achieve that, you can pass a QIIME parameters file to the workflow scripts to modify options to the
component scripts of the workflow. The parameters file is a text file with one parameter setting per line. Blank lines or
lines beginning with a ``#`` are ignored. A parameter setting is defined as ``script_name:parameter_name``, followed by
whitespace (space or tab), and then the value. The followings list the parameter files for the open reference, closed
reference, and _de novo_ OTU picking strategies.

### Open Reference OTUs ###
In an open-reference OTU picking process, reads are clustered against a reference sequence collection and any reads which
do not hit the reference sequence collection are subsequently clustered _de novo_. ``pick_open_reference_otus.py`` is the
primary interface for open-reference OTU picking in QIIME, and includes taxonomy assignment, sequence alignment, and
tree-building steps.

```python
"its99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_99/sh_taxonomy_qiime_ver8_99_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019"]

"its97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_97/sh_taxonomy_qiime_ver8_97_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019"]

"homd": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.qiime.taxonomy",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15"]

"gg99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus"]

"gg97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_97/97_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus"]

"sv99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_99/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S"]

"sv97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_97/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S"]
```

### Closed Reference OTUs ###
In a closed-reference OTU picking process, reads are clustered against a reference sequence collection and any reads which
do not hit a sequence in the reference sequence collection are excluded from downstream analyses.
``pick_closed_reference_otus.py`` is the primary interface for closed-reference OTU picking in QIIME. If the user provides
taxonomic assignments for sequences in the reference database,
those are assigned to OTUs.

```
"its99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_99/sh_taxonomy_qiime_ver8_99_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta"]

"its97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_97/sh_taxonomy_qiime_ver8_97_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019"]

"homd": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.qiime.taxonomy",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.qiime.taxonomy",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15"]

"gg99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus"]

"gg97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/Greengenes_97/97_otu_taxonomy.txt",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_97/97_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus"]

"sv99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/SILVA_99/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_99/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S"]

"sv97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/SILVA_97/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_97/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S"]
```

### _de novo_ Reference OTUs ###
In a _de novo_ OTU picking process, reads are clustered against one another without any external reference sequence
collection. ``pick_de_novo_otus.py`` is the primary interface for _de novo_ OTU picking in QIIME, and includes taxonomy
assignment, sequence alignment, and tree-building steps. A benefit of _de novo_ OTU picking is that all reads are
clustered. A drawback is that there is no existing support for running this in parallel in QIIME, so it can be too slow
to apply to large datasets (e.g., more than 10 million reads).

```python
"its99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_99/sh_taxonomy_qiime_ver8_99_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019"]

"its97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019",
    "pick_otus.py:pick_otus_reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/ITS_97/sh_taxonomy_qiime_ver8_97_02.02.2019.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019"]

"homd": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsumaclust",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.qiime.taxonomy",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15"]

"gg99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsumaclust",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus"]

"gg97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsumaclust",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_97/97_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_97/97_otus"]

"sv99": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsumaclust",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_99/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_99/otus_16S"]

"sv97": [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsumaclust",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/SILVA_97/majority_taxonomy_7_levels.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/SILVA_97/otus_16S"]
```

### Reference Databases ###
The pipeline currently supports several databases, namely [SILVA 97/99](http://www.arb-silva.de),
[Greengenes 97/99](https://greengenes.secondgenome.com), [Human Oral Microbiome Database (HOMD)](http://www.homd.org/) and
[ITS97/99](https://unite.ut.ee/). The following is the parameter file that specifies the locations of database for analysis.

```python
ref_db = {
    "homd": "/mnt/EFS/dbs/homd/HOMD_16S_rRNA_RefSeq_V15.1.p9.fasta",
    "gg99": "/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "gg97": "/mnt/EFS/dbs/Greengenes_97/97_otus.fasta",
    "sv99": "/mnt/EFS/dbs/SILVA_99/otus_16S.fasta",
    "sv97": "/mnt/EFS/dbs/SILVA_97/otus_16S.fasta",
    "its99": "/mnt/EFS/dbs/ITS_99/sh_refs_qiime_ver8_99_02.02.2019.fasta",
    "its97": "/mnt/EFS/dbs/ITS_97/sh_refs_qiime_ver8_97_02.02.2019.fasta"}
```

### PICRUSt Options ###
PICRUSt is designed to estimate the gene families contributed to a metagenome by bacteria or archaea identified using 16S
rRNA sequencing. Intermediate steps in this pipeline may also be of independent interest, as they allow for phylogenetic
prediction of organismal traits using reference examples (here applied to the problem of gene content prediction), and
correction for variable marker gene copy number. A common and heavily studied problem in phylogenetics involves estimating
the properties of ancestral organisms from living relatives (ancestral state reconstruction or ASR). PICRUSt extends these
methods to predict traits of living organisms that are unknown not because they belong to ancestral organisms, but because
they have simply not yet been studied. The main application of this Hidden State Prediction (HSP) in the core PICRUSt
workflow is to estimate the gene content of microorganisms for which no genome sequence is available, by using their
sequenced relatives as a reference.

```python
l2_param = [
    "summarize_taxa:md_identifier\t\"KEGG_Pathways\"",
    "summarize_taxa:absolute_abundance\tTrue",
    "summarize_taxa:level\t2"]

l3_param = [
    "summarize_taxa:md_identifier\t\"KEGG_Pathways\"",
    "summarize_taxa:absolute_abundance\tTrue",
    "summarize_taxa:level\t3"]

gg_param = [
    "pick_otus:enable_rev_strand_match\tTrue",
    "pick_otus:otu_picking_method\tsortmerna",
    "pick_otus:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus",
    "pick_otus:refseqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "alpha_diversity:metrics\tobserved_species,chao1,PD_whole_tree,shannon",
    "make_distance_boxplots:num_permutations\t0",
    "summarize_taxa:level\t2,3,4,5,6,7",
    "summarize_taxa:absolute_abundance\tTrue",
    "make_otu_table:taxonomy\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:id_to_taxonomy_fp\t/mnt/EFS/dbs/Greengenes_99/99_otu_taxonomy.txt",
    "assign_taxonomy:reference_seqs_fp\t/mnt/EFS/dbs/Greengenes_99/99_otus.fasta",
    "assign_taxonomy:assignment_method\tsortmerna",
    "assign_taxonomy:sortmerna_db\t/mnt/EFS/dbs/Greengenes_99/99_otus"]
```

## Pipeline Output Files/Folders ##
- ``split_lib_out``: Demultiplex and quality filter (usually at Phred > Q20) of paired Illumina reads. For more information,
please visit QIIME1 online documentation.
- ``otus``: The OTU picking step assigns similar sequences to operational taxonomic units, or OTUs, by clustering sequences
based on a user-defined similarity threshold. Sequences which are similar at or above the threshold level are taken to
represent the presence of a taxonomic unit (e.g., a genus, when the similarity threshold is set at 0.94) in the sequence
collection. The final OTU table is summarized in a BIOM file, e.g., ``otu_table_mc2_w_tax_no_pynast_failures.biom`` for open
reference OTU picking strategy and ``otu_table.biom`` for closed and _de novo_. These BIOM files are used for the downstream
analysis.
- ``core_dirersity``: The diversity analysis generate figures and tables according to the metadata in the mapping file. The
analysis includes alpha diversity, beta diversity, distance boxplots, comparison of alpha diversity, and quantitative
analysis of group differences. The output from core diversity is conveniently summarized in a HTML file (``index.html``). The
HTML file lists all the quantitative analyses such as alpha and beta diversity, group significance, and the corresponding
figures and summaries. <b>Note</b>: The QIIME1 pipeline no longer generates the taxonomy summary plots
(``plot_taxa_summary.py``). For microbial communities with complex profiles, it can take a few weeks to generate the taxonomy
summary plots.
- ``graphs``: The visualization pipeline runs the plotly graphs and generates an interactive heatmap.
- ``otu_summary_table.txt``: The file lists the number of taxonomically assigned reads for each sample in the dataset.
- ``otu_table.txt``: A tab-separated text file of sequence variant counts and taxonomic assignment at the genus level.
- ``otu_picrust`` (optional): Closed reference OTU picking using QIIME1 with Greengenes 99 database.
- ``PICRUST_data`` (optional): This directory contains the final metagenome functional predictions with figures and bar
plots.

# Pipeline Steps and Modules #
The QIIME pipelines share the modules that perform join paired-end reads, taxonomic assignment and core diversity analysis.
This was done to simplify the implementation and to make the pipelines more robust. In essence, the implementation of the
pipelines embraces modular design. A module is defined as the unique and addressable components of the software which can be
solved and modified independently without disturbing or interfering other modules of the software. Effective modular design
can be achieved if the partitioned modules are separately solvable, modifiable as well as compilable. Here separate
compilable modules means that after making changes in a module there is no need of recompiling or modifying the whole
software system. In order to build a software with effective modular design there is a factor <q>functional independence</q>
which comes into play. Functional independence implies that a function is atomic in nature so that it performs only a single
task of the software without or with least interaction with other modules. Since the functionality of the software have been
broken down into atomic levels, thus developers get a clear requirement of each and every functions and hence designing of
the software becomes easy and error free. As the modules are independent they have limited or almost no dependency on other
modules. So, making changes in a module without affecting the whole system is possible in this approach. Error propagation
from one module to another and further in whole system can be neglected and it saves time during testing and debugging.

## Join Paired-End Reads ##
This module, ``join_read.py``, joins the paired-end reads and runs the algorithm with all available processor cores. It wraps
the script, ``join_paired_ends.py``, which takes forward and reverse Illumina reads and joins them using the method chosen.
The script will optionally create an updated index reads file containing index reads for the surviving joined paired end
reads. If the option to write an updated index file is chosen, be sure that the order and header format of the index reads is
the same as the order and header format of reads in the files that will be joined (this is the default for reads generated on
the Illumina instruments).

To run the unit test of the join paired-end module, type:
```bash
join_read.py --map map_file.txt --input input_dir --output output_dir --overlap 10 --max_diff 25
```

## Module Options ###
- ``--input``: input folder.
- ``--ouput``: output folder.
- ``--map_file``: mapping file that describes the metadata associated with samples. The mapping file should contain all of
the information about the samples necessary to perform the data analysis. In general, the mapping file should contain the
name of each sample, the barcode sequence used for each sample, the linker/primer sequence used to amplify the sample, and
a description column. One should also include in the mapping file any metadata that relates to the samples (for instance,
health status or sampling site) and any additional information relating to specific samples that may be useful to have at
hand when considering outliers (for example, what medications a patient was taking at time of sampling). See
[QIIME documentation](http://qiime.org/documentation/file_formats.html) for the format.
- ``--overlap``: minimum allowed overlap in bases required to join pairs. Not applicable to single-end samples. The
default value is 10.
- ``--max_diff``: maximum allowed percent differences within overlap. Not applicable to single-end samples. The default
value is 25.

## Taxonomic Assignment ##
The module, ``pick_otu.py``, performs OTU picking with the strategy _de novo_, closed reference, and open reference. All of
the sequences from each sample will be clustered into Operational Taxonomic Units (OTUs) based on their sequence similarity.
There are seven databases to choose for taxonomic assignment, namely HOMD (Human Oral Microbiome Database; v15.1), SILVA 97
and SILVA 99 (v132), Greengenes 97 and Greengenes 99 (v13.8), and ITS97 and ITS99 (UNITE+INSDC 18.11.2018). The ITS databases
have been updated to the latest version with 817,130 sequences. The sequences in the UNITE databases target the formal fungal
barcode, the nuclear ribosomal internal transcribed spacer (ITS) region. UNITE uses the NCBI taxonomy classification with
modifications from Index Fungorum. The sequences in these databases are clustered at 97 and 99% identity. QIIME provides
three high-level protocols for OTU picking. These are described as _de novo_, closed-reference, and open-reference OTU
picking:

- _de novo_ (``pick_de_novo_otus.py``): Reads are clustered against one another without an external reference sequence
collection. Useful for studying populations where there is poor characterization of existing data. In addition to clustering,
the pipeline also performs taxonomy assignment, sequence alignment, and tree-building steps.
- closed reference (``pick_closed_reference_otus.py``): Reads are clustered against a reference sequence collection and any
reads which do not hit a sequence in the reference sequence collection are excluded. Useful when existing taxa are well
characterized and one is not interested in novel ones. Taxonomic assignments in the reference database are assigned to the
OTUs.
- open reference (``pick_open_reference_otus.py``): Reads are clustered against a reference sequence collection and any reads
which do not hit the reference collection are clustered de novo. Applicable when existing taxa are well characterized and one
also wishes to discover novel ones.

For a more detailed discussion of these OTU picking protocols, please see
[Rideout et al. (2014)](https://peerj.com/articles/545/). It is important to note that QIIME does not actually implement OTU
picking algorithms, but rather wraps external OTU clustering tools. There are a number of OTU clustering tools available
through QIIME’s workflows, including open source (e.g., SortMeRNA, SUMACLUST, and swarm) and closed source tools (e.g.,
``uclust`` and ``usearch``). [SortMeRNA](https://bioinfo.lifl.fr/RNA/sortmerna/) is the default OTU clustering tool used in
the pipelines. SortMeRNA is a program tool for filtering, mapping and OTU-picking NGS reads in metatranscriptomic and
metagenomic data. The core algorithm is based on approximate seeds and allows for fast and sensitive analyses of nucleotide
sequences. The main application of SortMeRNA is filtering ribosomal RNA from metatranscriptomic data. For more information,
please see: [E Kopylova, et al. (2012)](bioinformatics.oxfordjournals.org/content/28/24/3211).

To run the unit test of the core diversity module, type:
```bash
pick_otu.py --input input_dir --output output_dir --phred 19 --bad_run 3 --max_n 0 --offset 33 --strategy "open" --ref_db "gg97"
```

### Module Options ###
- ``--input``: input folder.
- ``--ouput``: output folder.
- ``--phred``: maximum unacceptable [Phred quality score](https://www.phrap.com/phred/). The Phred quality score is a measure
of the quality of the identification of the nucleobases generated by the sequencing platforms such as Illumina and 454.
Minimum Phred quality score of 19 for Q20 (1 in 100 of incorrect base call) or better is recommended. The default is 19.
- ``--max_bad_run``: maximum number of consecutive low quality base calls permitted before truncation. For V1-V3, a minimum
value of 10 is recommended. The default value is 3.
- ``--max_n``: maximum number of degenerate base (N) characters allowed in a sequence to retain it. This is applied after
quality trimming, and is total over combined paired end reads if applicable maximum number of degenerate bases. The default
value is 0.
- ``--offset``: the ASCII offset to use when decoding [Phred scores]((https://www.phrap.com/phred/)) (either 33 or 64).
The default is 33. Phred Q scores are often represented as ASCII characters of base 33 and 64. The default is base 33.
- ``--strategy``: OTU picking strategy. The options are ``open``, ``closed``, and ``de_novo``.
- ``--ref_db``: reference database for OTU picking. The options are ``homd``, ``sv97``, ``sv99``, ``gg97``, ``gg99``,
``its97``, and ``its99``.

## Analysis of Core Diversity ##
This module wraps the QIIME script, ``core_diversity_analyses.py``, for core diversity analysis on the samples. The QIIME
core diversity script plugs several analyses together to form a basic workflow beginning with a BIOM table, mapping file,
and optional phylogenetic tree. The included scripts are those run by the workflow scripts ``alpha_rarefaction.py``,
``beta_diversity_through_plots.py``, ``summarize_taxa_through_plots.py``, plus the (non-workflow) scripts
``make_distance_boxplots.py``, ``compare_alpha_diversity.py``, and ``group_significance.py``. To update parameters to the
workflow scripts, you should pass the same parameters file that you would pass if calling the workflow script directly.
Additionally, a table summary is generated by running the ``biom summarize-table`` command (part of the
[BIOM file format](http://biom-format.org/) package). To update parameters to this command, your parameters file should use
``biom-summarize-table`` as the script name.

To run the unit test for this module, type:
```bash
diversity.py --map map_file.txt --input input_dir --output output_dir --depth 0 --biom otu.biom --tree homd
```

### Module Options ###
- ``--input``: input folder.
- ``--ouput``: output folder.
- ``--depth``: sampling depth for downstream analysis. The number of counts for filtering and subsampling the OTU table for
downstream analysis. Samples which have counts below this value will be removed from the downstream analysis. The counts of
the remaining samples will be subsampled to this value. If not specified, it will be calculated automatically. The default
value is 0.
- ``--biom``: input BIOM file generated from the OTU picking step.
- ``--tree``: phylogenetic tree associatd with the reference databases for downstream analysis. The options are ``sv97``,
``sv99``, ``gg97``, ``gg99``, and ``homd``. Note: reference databases, ITS97 and ITS99, do not come with phylogenetic trees.

## PICRUSt (Optional) ##
PICRUSt is designed to estimate the gene families contributed to a metagenome by bacteria or archaea identified using 16S
rRNA sequencing. Intermediate steps in this pipeline may also be of independent interest, as they allow for phylogenetic
prediction of organismal traits using reference examples (here applied to the problem of gene content prediction), and
correction for variable marker gene copy number.

To run the unit test for the PICRUSt module, type the command:
```bash
picrust.py --input input_dir --output output_dir --fasta input.fasta
```

### Module Options ###
- ``--input``: input folder.
- ``--ouput``: output folder.
- ``--fasta``: FASTA for functional annotation. The FASTA file is the output from the closed reference OTU picking with the
Greengenes 99 database.

# References #
* MGI Langille, J Zaneveld, JG Caporaso, D McDonald, D Knights, JA Reyes, JC Clemente, DE Burkepile, RL Vega Thurber,
R Knight, RG Beiko and C Huttenhower (2013) Predictive functional profiling of microbial communities using 16S rRNA marker
gene sequences. *Nature Biotechnology*. [doi: 10.1038/nbt.2676](https://www.nature.com/articles/nbt.2676).

* JG Caporaso, J Kuczynski, J Stombaugh, K Bittinger, FD Bushman, EK Costello, N Fierer, AG Pena, JK Goodrich, JI Gordon,
GA Huttley, ST Kelley, D Knights, JE Koenig, RE Ley, CA Lozupone, D McDonald, BD Muegge, M Pirrung, J Reeder, JR Sevinsky,
PJ Turnbaugh, WA Walters, J Widmann, T Yatsunenko, J Zaneveld and R Knight (2010) QIIME allows analysis of high-throughput
community sequencing data. *Nature Methods*. [doi:10.1038/nmeth.f.303](https://www.nature.com/articles/nmeth.f.303).

* E Kopylova, L Noé and H Touzet (2012) SortMeRNA: Fast and accurate filtering of ribosomal RNAs in metatranscriptomic data,
*Bioinformatics*. [doi:10.1093/bioinformatics/bts611](bioinformatics.oxfordjournals.org/content/28/24/3211).

* TZ DeSantis, P Hugenholtz, N Larsen, M Rojas, EL Brodie, K Keller, T Huber, D Dalevi, P Hu, GL AndersenGreengenes (2006)
a Chimera-Checked 16S rRNA Gene Database and Workbench Compatible with ARB. *Applied and Environmental Microbiology*.
[doi:10.1128/AEM.03006-05](https://aem.asm.org/content/72/7/5069).

* C Quast, E Pruesse, P Yilmaz, J Gerken, T Schweer, P Yarza, J Peplies, FO Glöckner (2013) The SILVA ribosomal RNA gene
database project: improved data processing and web-based tools. *Nucleic Acids Research*, 41(D1), D590-D596.
[doi:10.1093/nar/gks1219](https://academic.oup.com/nar/article/41/D1/D590/1069277).


*Last updated on March 18, 2020*.
